Update add/remove general layers and libraries, handle meta-rz-codecs, Panfrost graphic switching

phatnguyenxh opened the pull request:
I have updated the code to replace the add/remove meta-rz-features logic with a more general approach that covers layers under the build target directory (excluding the layers that are defined in the bblayers.conf.sample under meta-renesas). I also handled the inclusion of meta-rz-codecs files in meta-renesas for core-image-weston by adding a new conditional flag to control include/core-image-renesas-mmp.inc. The code also manages adding/removing libraries and switching between Panfrost graphics and no graphics.

phatnguyenxh committed:
Update add/remove general layers and libraries, handle meta-rz-codecs, Panfrost graphics switching

tiennguyenuh commented on rz-cmn-srp/config.json:
```diff
@@ -35,5 +35,22 @@
 		"machine": "rz-cmn",
 		"image": "core-image-weston"
+	},
+	"features": {
+        "layers": {
+            "add": [
+                "meta-openembedded/meta-perl",
+                "meta-rz-features/meta-rz-codecs"
+            ],
+            "remove": [
+                "meta-browser/meta-chromium",
+                "meta-clang"
+            ]
+        },
+		"libraries": {
+			"add": [],
+			"remove": []
+		},
+		"gpu": "none"
```
When you add new items or configurations to the config file, please make sure to update the README as well.

phatnguyenxh replied:
I have updated the README.md

tiennguyenuh commented on rz-cmn-srp/rz_builder.sh:
```diff
@@ -723,12 +725,217 @@ setup() {
+            elif ! grep -q '^CONFIG_DRM_PANFROST=y' "${DEFCONFIG}"; then
+                echo 'CONFIG_DRM_PANFROST=y' >> "${DEFCONFIG}"
+            fi
+            echo 'CONFIG_DRM_PANFROST=y' > "${PANFROST_CFG}"
```
So you’ve overridden the entire panfrost.cfg with CONFIG_DRM_PANFROST=y? That’s not a good idea, to be honest, and it’s not correct as well.

phatnguyenxh replied:
I have updated the meta-renesas to adapt a conditional flag and change the method to toggle that flag in the script

tiennguyenuh commented on rz-cmn-srp/rz_builder.sh:
```diff
@@ -723,12 +725,217 @@ setup() {
+    fi
+
+    # If this is a real layer (has conf/layer.conf) add it
+    if [ -f "${layer_path}/conf/layer.conf" ]; then
```
If it isn’t a layer, just add an else clause.

phatnguyenxh replied:
I have added the else clause for the conditions in add_layer() and remove_layer()

tiennguyenuh commented on rz-cmn-srp/rz_builder.sh:
```diff
@@ -723,12 +725,217 @@ setup() {
+    fi
+
+    # If this is a folder that contains sub-layers, add every child layer under it
+    if [ -d "${layer_path}" ]; then
```
If it isn’t a path, just add an else clause and remove the last log_warning

tiennguyenuh commented on rz-cmn-srp/rz_builder.sh:
```diff
@@ -723,12 +725,217 @@ setup() {
+    log_warning "Path does not exist or is not a layer: ${layer}"
+}
+
+remove_layer() {
```
Same comment as add_layer function

tiennguyenuh commented on rz-cmn-srp/rz_builder.sh:
```diff
@@ -723,12 +725,217 @@ setup() {
+    local layer_path
+
+    layer_path="${RZ_TARGET_DIR}/${layer}"
+    # Control flag to handle the dependencies of meta-rz-codecs in meta-renesas when building core-image-weston
```
It’s not just core-image-weston; we have several images that use the codec, so please use a more general name instead.

phatnguyenxh replied:
I have rename the comments and remove the core-image-weston mentioned

tiennguyenuh commented on rz-cmn-srp/rz_builder.sh:
```diff
@@ -814,6 +1021,17 @@ build() {
+    apply_add_remove_layers
+	apply_gpu_feature
+	apply_libraries
+
```
I think you can move these into the setup_conf function, since both build and build-sdk invoke it.

phatnguyenxh replied:
I have moved it into the setup_conf

tiennguyenuh commented on rz-cmn-srp/config.json:
```diff
@@ -35,5 +35,22 @@
+			"add": [],
+			"remove": []
+		},
+		"gpu": "none"
```
You can use `panfrost` as default

tiennguyenuh commented on rz-cmn-srp/rz_builder.sh:
```diff
@@ -723,12 +725,217 @@ setup() {
+
+apply_gpu_feature() {
+	LAYERDIR="${RZ_TARGET_DIR}/meta-renesas"
+    local DEFCONFIG="${LAYERDIR}/recipes-kernel/linux/rz-cmn/common/renesas_defconfig"
```
Check only the panfrost config in panfrost.cfg. If panfrost is accidentally included in renesas_defconfig, remove it in the meta-renesas PR; there’s no need to touch renesas_defconfig here.

phatnguyenxh replied:
I have updated the meta-renesas to adapt a conditional flag and change the method to toggle that flag in the script

tiennguyenuh replied:
Have you raised a PR for meta-renesas? I can’t find any.

phatnguyenxh replied:
I have created the PR: https://github.com/vudangRVC/meta-renesas-sst/pull/5

tiennguyenuh commented on rz-cmn-srp/rz_builder.sh:
```diff
@@ -723,12 +725,217 @@ setup() {
+    [ -f "${lc}" ] || touch "${lc}"
+
+    # Removing any previous lines to avoid duplicates
+    sed -i '\|^IMAGE_INSTALL:append = " \${USER_IMAGE_ADD}"$|d' "${lc}"
```
Your sed lines delete only the exact literal ${USER_IMAGE_ADD} form. If the line previously contained an expanded value, it won’t match and duplicates will accumulate. For example:

```
IMAGE_INSTALL:append = " foo foo1 foo2 add-package"
```

phatnguyenxh replied:
The method worked fine, I made the function explaination unclearly. The conf_clean_libraries() only creates template and the parsing from config.json will be handled by the conf_set_variable() under apply_libraries(), so the libraries can be parsed cleanly and avoid duplicates
`conf_set_variable 'USER_IMAGE_ADD' "${ADD_LIBS}"`
 `conf_set_variable 'USER_PACKAGE_EXCLUDE' "${REMOVE_LIBS}"`

tiennguyenuh commented on rz-cmn-srp/rz_builder.sh:
```diff
@@ -723,12 +725,217 @@ setup() {
+            echo '# CONFIG_DRM_PANFROST is not set' > "${PANFROST_CFG}"
+            ;;
+        *)
+            log_warn "Unknown GPU mode '${GPU_MODE}', setting back to none"
```
log_warn doesnt exist

phatnguyenxh replied:
I have noticed and changed

tiennguyenuh reviewed (commented):
@phatnguyenxh, please make sure to test the script in a POSIX environment.

phatnguyenxh committed:
rz-cmn: Update the apply_gpu_feature to control flag in meta-renesas, add_layer, remove_layer and minor comments

phatnguyenxh committed:
rz-cmn: Update README.md including add/remove layer, panfrost features

phatnguyenxh committed:
rz-cmn: Update the functions move from build() to setup_conf, change minimal comments

tiennguyenuh commented on rz-cmn-srp/README.md:
````diff
@@ -5,45 +5,96 @@ This directory contains automated build scripts and resources for performing Yoc
 ## Hierarchy
 
 ```
-$ tree -L 3
````
Please use tree -L 3, this root readme is just for general overview.

phatnguyenxh replied:
I have fixed it to tree -L 3, the tree is diffrent because I changed the path of meta-rz-features to meta-rz-features/meta-rz-codecs in `files_to_add` and `patches`

tiennguyenuh replied:
I meant just using tree level 3, not level 4.

tiennguyenuh commented on rz-cmn-srp/README.md:
```diff
@@ -101,6 +152,32 @@ It also lists the available `machine` types
     - `machine` : Specify the default machine chosen when no machine is passed as arguement.
     - `image` : Specify the default image to build where none is specified.
 
+- **features**: Quick toggles applied by the build scripts before BitBake. Lets you adjust layers, image packages, and GPU settings without editing code.
```
Always use a neutral tone. When writing documentation, avoid words such as ‘we,’ ‘you,’ and ‘me, etc.

phatnguyenxh replied:
I have changed the words

tiennguyenuh commented on rz-cmn-srp/README.md:
```diff
@@ -101,6 +152,32 @@ It also lists the available `machine` types
+- **features**: Quick toggles applied by the build scripts before BitBake. Lets you adjust layers, image packages, and GPU settings without editing code.
+    - `layers`: Control Yocto layers. Supports either a specific layer path (has `conf/layer.conf`) or a folder of layers (adds/removes all valid sub-layers under it).
+    - `libraries`: Control image packages. `add` appends packages, `remove` excludes packages (and marks as bad recommendations).
+    - `gpu`: Choose the graphic mode between Panfrost or no support. Use `"none"` or `"panfrost"` to toggle the configuration flag in kernel.
```
Please list the available gpu options and briefly explain each. There is also a mali option, but it is not currently supported, add a remark to note this.

phatnguyenxh replied:
I have added the list of gpu and rewrote the explaination more clearly

tiennguyenuh commented on rz-cmn-srp/README.md:
```diff
@@ -101,6 +152,32 @@ It also lists the available `machine` types
     - `image` : Specify the default image to build where none is specified.
 
+- **features**: Quick toggles applied by the build scripts before BitBake. Lets you adjust layers, image packages, and GPU settings without editing code.
+    - `layers`: Control Yocto layers. Supports either a specific layer path (has `conf/layer.conf`) or a folder of layers (adds/removes all valid sub-layers under it).
```
I think its not clear, please clarify that layers adds/removes entries of meta-layer in BBLAYERS

phatnguyenxh replied:
I have rephrased this and clarified the sentences

tiennguyenuh commented on rz-cmn-srp/rz_builder.sh:
```diff
@@ -723,6 +737,174 @@ setup() {
+    log_info "GPU mode: ${GPU_MODE}"
+
+	# Control RZ_FEATURE_PANFROST in meta-renesas
+    case "${GPU_MODE}" in
```
Add a line before the case to disable Panfrost by default, avoiding duplicated logic.

- If panfrost -> enable it.
- If mali -> note that it is not currently supported.
- If none (or empty) -> no action.

phatnguyenxh replied:
I have changed the logic to match that flow

tiennguyenuh commented on rz-cmn-srp/rz_builder.sh:
```diff
@@ -723,6 +737,174 @@ setup() {
+	# If this is a folder that contains sub-layers, add every child layer under it
+    elif [ -d "${layer_path}" ]; then
+        local match
+        if ! match=$(compgen -G "${layer_path}/*/conf/layer.conf"); then
```
You’re using compgen, is that POSIX compliant?

phatnguyenxh replied:
Sorry, I have changed the command and logic to be compatible with POSIX

phatnguyenxh committed:
rz-cmn: Update logic and README.md
Changes include:
    - rz_builder.sh: Update logic flow of apply_gpu_feature.
    - README.md: Rephrase the words

Signed-off-by: Phat Nguyen <phat.nguyen.xh@bp.renesas.com>

phatnguyenxh committed:
rz-cmn: Reupdate function logic and syntax
Changes include:
    - rz_builder.sh: Corrected some logic and syntax
